File: src/main/java/com/tobe/healthy/notification/domain/Notification.java
package com.tobe.healthy.notification.domain;

import static jakarta.persistence.EnumType.*;

import org.hibernate.annotations.DynamicUpdate;

import com.tobe.healthy.common.BaseTimeEntity;
import com.tobe.healthy.member.domain.Member;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;

@Entity
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@DynamicUpdate
public class Notification extends BaseTimeEntity<Notification, Long> {

	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	@Column(name = "notification_id")
	private Long id;

	private Long studentId;

	private String studentName;

	private String clickUrl;

	private String title;

	private String content;

	@Enumerated(STRING)
	private NotificationCategory notificationCategory;

	@Enumerated(STRING)
	private NotificationType notificationType;

	@Enumerated(STRING)
	private NotificationSenderType senderType;

	@ManyToOne(fetch = FetchType.LAZY)
	@JoinColumn(name = "receiver_id")
	private Member receiver;

	private Long targetId;

	private boolean isRead;

	private Notification(String title, String content, NotificationCategory notificationCategory,
		NotificationType notificationType, Member receiver, Long targetId,
		String clickUrl, Long studentId, String studentName) {
		this.title = title;
		this.content = content;
		this.notificationCategory = notificationCategory;
		this.notificationType = notificationType;
		this.senderType = NotificationSenderType.SYSTEM;
		this.receiver = receiver;
		this.targetId = targetId;
		this.clickUrl = clickUrl;
		this.studentId = studentId;
		this.studentName = studentName;
		this.isRead = false;
	}

	public static Notification create(String title, String content,
		NotificationCategory notificationCategory,
		NotificationType notificationType,
		Member receiver, Long targetId,
		String clickUrl, Long studentId, String studentName) {
		return new Notification(title, content, notificationCategory, notificationType,
			receiver, targetId, clickUrl, studentId, studentName);
	}

	public void updateNotificationStatus() {
		this.isRead = true;
	}
}

File: src/main/kotlin/com/tobe/healthy/notification/domain/entity/Notification.kt
package com.tobe.healthy.notification.domain.entity

import com.tobe.healthy.common.BaseTimeEntity
import com.tobe.healthy.member.domain.entity.Member
import com.tobe.healthy.notification.domain.entity.NotificationSenderType.SYSTEM
import jakarta.persistence.*
import jakarta.persistence.EnumType.STRING
import jakarta.persistence.FetchType.LAZY
import jakarta.persistence.GenerationType.IDENTITY
import org.hibernate.annotations.DynamicUpdate

@Entity
@DynamicUpdate
class Notification(

    val studentId: Long? = null,

    val studentName: String? = null,

    val clickUrl: String? = null,

    val title: String,

    val content: String,

    @Enumerated(STRING)
    val notificationCategory: NotificationCategory,

    @Enumerated(STRING)
    val notificationType: NotificationType,

    @Enumerated(STRING)
    val senderType: NotificationSenderType = SYSTEM,

    @ManyToOne(fetch = LAZY)
    @JoinColumn(name = "receiver_id")
    val receiver: Member? = null,

    val targetId: Long? = null,

    var isRead: Boolean = false,

    @Id
    @GeneratedValue(strategy = IDENTITY)
    @Column(name = "notification_id")
    val id: Long? = null

) : BaseTimeEntity<Notification, Long>() {

    fun updateNotificationStatus() {
        this.isRead = true
    }

    companion object {
        fun create(title: String, content: String, notificationCategory: NotificationCategory, notificationType: NotificationType, receiver: Member, targetId: Long? = null, clickUrl: String? = null, studentId: Long? = null, studentName: String? = null
        ): Notification {
            return Notification(
                title = title,
                content = content,
                notificationCategory = notificationCategory,
                notificationType = notificationType,
                receiver = receiver,
                targetId = targetId,
                clickUrl = clickUrl,
                studentId = studentId,
                studentName = studentName
            )
        }
    }
}